Lib/idlelib/idle_test/test_pyparse.py

```diff
@@ -0,0 +1,521 @@
+        self.assertEqual(self.mapping[ord('a')], self.default)
+
+    def test__len__(self):
+        self.assertEqual(len(self.mapping), len(self.preserve_dict))
+
+    def test__iter__(self):
+        count = 0
+        for key, value in self.mapping.items():
+            self.assertIn(key, self.preserve_dict)
+            count += 1
+        self.assertEqual(count, len(self.mapping))
+
+    def test_get(self):
+        self.assertEqual(self.mapping.get(ord('\t')), ord('\t'))
+        self.assertEqual(self.mapping.get('a'), self.default)
+        # Default is a parameter, but it isn't used.
+        self.assertEqual(self.mapping.get('a', default=500), self.default)
+
+
+class PyParseTest(unittest.TestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        cls.parser = pyparse.Parser(indentwidth=4, tabwidth=4)
+
+    @classmethod
+    def tearDownClass(cls):
+        del cls.parser
+
+    def test_init(self):
+        self.assertEqual(self.parser.indentwidth, 4)
+        self.assertEqual(self.parser.tabwidth, 4)
+
+    def test_set_str(self):
+        eq = self.assertEqual
+        p = self.parser
+        setstr = p.set_str
+
+        # Not empty and doesn't end with newline.
+        with self.assertRaises(AssertionError):
+            setstr('a')
+
+        tests = ('',
+                 'a\n')
+
+        for string in tests:
+            with self.subTest(string=string):
+                setstr(string)
+                eq(p.str, string)
+                eq(p.study_level, 0)
+
+    def test_find_good_parse_start(self):
+        eq = self.assertEqual
+        p = self.parser
+        setstr = p.set_str
+        start = p.find_good_parse_start
+
+        # Split def across lines.
+        setstr('"""This is a module docstring"""\n'
+               'class C():\n'
+               '    def __init__(self, a,\n'
+               '                 b=True):\n'
+               '        pass\n'
+               )
+
+        # No value sent for is_char_in_string().
+        self.assertIsNone(start())
+
+        # Make text look like a string.  This returns pos as the start
+        # position, but it's set to None.
+        self.assertIsNone(start(is_char_in_string=lambda index: True))
+
+        # Make all text look like it's not in a string.  This means that it
+        # found a good start position.
+        eq(start(is_char_in_string=lambda index: False), 44)
+
+        # If the beginning of the def line is not in a string, then it
+        # returns that as the index.
+        eq(start(is_char_in_string=lambda index: index > 44), 44)
+        # If the beginning of the def line is in a string, then it
+        # looks for a previous index.
+        eq(start(is_char_in_string=lambda index: index >= 44), 33)
+        # If everything before the 'def' is in a string, then returns None.
+        # The non-continuation def line returns 44 (see below).
+        eq(start(is_char_in_string=lambda index: index < 44), None)
```

Comment:
In the `tries` for loop, the first thing it does is `rfind` for ':\n', so in this example, that's the end of the def statement (pos 109).  It then finds the start of the line as `rfind('\n')`, which does not return the start of the line that actually has the `def` on it (it returns pos 70 instead of 44).  `_synchre` doesn't find a keyword, so it gets ready to loop again.  On the next loop, it again does `rfind` for ':\n', which is the end of the `class` statement.  But the class statement is in a string (pos 33), so that's no good.  This means it enters `if pos is None:`.  Since that invokes _synchre from the beginning of the string, it again finds `class` and again rejects it.

One goal of this routine seems to be to find a better start point as quickly as possible.  It sacrifices 'being perfect' for just trying a few (5) times.  So, I don't know if this is a big deal that it doesn't find it.  The downstream effect is that _study2 (and to a lesser extent _study1) might have to process more of the string instead of the 'best guess' of the tail of the string.

Reply:
Thanks.  I missed the point that def is only found in the 'while 1' loop after backing up to class and searching forward.  See msg312525 of bpo-32880 for more thoughts.